staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -158,25 +231,91 @@ func (q *delayingType) ShutDown() {
 	})
 }
 
-// AddAfter adds the given item to the work queue after the given delay
+// Add follows the same behaviour as Add() on Interface(), except that it also
+// removes a duplicate 'item' from the waiting queue, if it exists.
+// We can say that the immediate add is pre-empting the same add that was waiting.
+// Calls to Add() are synchronous. i.e. the item will have been added or dropped before return.
+func (q *delayingType) Add(item interface{}) {
```

Comment:
Overriding the queue `Add` is going to break people who are depending on the current behaviour. 

In k8s this will break or slow down availability checking for most of the workload controllers in certain scenarios.

- in [replica sets](https://github.com/kubernetes/kubernetes/blob/3f823c0daa002158b12bfb2d53bcfe433516659d/pkg/controller/replicaset/replica_set.go#L464) and thus deployments could cause skipping availability checks and stop progression of deployments
- in stateful sets and daemon sets, it could cause  slower checking of availability

Basically for everybody who is using the delaying queue as a remainder to do something. 
And if meanwhile irrelevant change is observed, the reminder will be lost.

It is not a great that we or potentially other people depend on this behaviour. But I think it would be preferable to drop the breaking changes from this PR  and do some outreach first so people can get accustomed to this change. And once considered safe, introduce this change later.

I am planning to fix our controllers and I am already proposing a fix for replica sets (plus deployments) in https://github.com/kubernetes/kubernetes/pull/113605, but I need this PR functionality to proceed (mainly `TakeLonger`)


Nevertheless I recognize, that this change is desirable for most of the use cases to stop processing items multiple times (https://github.com/kubernetes/kubernetes/issues/110642)

Reply:
> Overriding the queue Add is going to break people who are depending on the current behaviour

If this were true I would expect some e2e or other test failures. Can you link to a specific code location that this will break?

Reply:
Our e2e's are not exactly great in this area (availability checking). That is why this bug also exists: https://github.com/kubernetes/kubernetes/issues/108266 which I am trying to fix and is validate with these tests https://github.com/kubernetes/kubernetes/pull/113605/files#diff-ab640ef4976871fe1ead6a05a01c9a0755b8e4b3996473f27613e09956f2183cR901. But they are focusing on that bug.

I can create a test that would break this in a more generic scenario. Basically we are scheduling reminders here https://github.com/kubernetes/kubernetes/blob/3f823c0daa002158b12bfb2d53bcfe433516659d/pkg/controller/replicaset/replica_set.go#L464, which will be undercut by rs informer for any change. And yes the current replica set implemention is imperfect.

Reply:
in sts and ds we have a better fallback here 		dsc.enqueueDaemonSetAfter(ds, time.Duration(ds.Spec.MinReadySeconds)*time.Second). But it will also cause delays in availability. For example with 30s minReadySeconds  you might end up with availability being reported in 45s. In this case the reminder got interrupted after 15s by a sts/ds informer.

Reply:
How is that going to break? Add() is still going to beat any AddAfter with a non-zero delay through the queue.

Reply:
If I understand what you're saying, the current implementation also has this problem and this PR won't make it any different or worse?

Reply:
> ```
> // Add follows the same behaviour as Add() on Interface(), except that it also
> // removes a duplicate 'item' from the waiting queue, if it exists.
> ```

It is going to make it worse, because it will cancel the reminder which we need for checking the availability. The other `Add` might be important for some other update/reconciilation, but not for the availability check.

Reply:
I see. Maybe we need a way to make some queue entries sticky, sigh. I don't think that's the normal case.

Reply:
Our use case should be covered by `AddAfterWithOptions` with `AddAfterOptions{Existing: TakeLonger}` and we can switch the controllers to use that after this gets in. It is certainly not a normal case, but I am worried there might be  people who are relaying on this behaviour like we do now.

Reply:
^ covered together with a separate queue which will remove the need for stickyness https://github.com/kubernetes/kubernetes/pull/113605/files#diff-f658ce5cda23ed9920c4ee4eaaee968fb5a4dbd4a0ccc7f478db399f2b182380R118

Reply:
Since the queue already today doesn't offer any guarantees about things being sticky (if someone uses AddAfter with a shorter duration), yeah a separate queue sounds safe today and into the future.

If we do add a sticky option, it should be the job of the code putting the sticky entry into the queue to set that, not the job of all the other Add locations to avoid stepping on it, IMO.

Reply:
In case you decide to go with this `Add` change now, we need to refactor our controllers before this PR gets in. Othewise we can do the refactoring after and react to this PR.

Nevertheless, I  think it would be preferable to make people aware.

Reply:
If the delaying_queue allowed multiple queuings of the same item 'x' then I would certainly see this as a more generic interpretation of a delaying queue than the existing implementation.  Then I think the pre-empting and de-dup behaviours of the queue would be an opt-in thing (however, it is live with this today and this is tested behaviour so such a change will definitely have more impact on clients - unless we move it exclusively to new methods or a new type even).